components/org.wso2.carbon.identity.auth.valve/src/main/java/org/wso2/carbon/identity/auth/valve/util/APIErrorResponseHandler.java

```diff
@@ -0,0 +1,160 @@
+package org.wso2.carbon.identity.auth.valve.util;
+
+import org.apache.catalina.connector.Response;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.json.JSONArray;
+import org.json.JSONObject;
+import org.wso2.carbon.identity.auth.service.AuthenticationContext;
+
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+
+import org.apache.log4j.MDC;
+
+public class APIErrorResponseHandler {
+
+    private static final Log log = LogFactory.getLog(APIErrorResponseHandler.class);
+
+    private static final String AUTH_HEADER_NAME = "WWW-Authenticate";
+    private static final String CORRELATION_ID_MDC = "Correlation-ID";
+
+    public static void handleErrorResponse(AuthenticationContext authenticationContext, Response response, int error,
+                                            Exception e) throws IOException {
+
+        if (log.isDebugEnabled() && e != null) {
+            log.debug("Authentication Error ", e);
+        }
+        StringBuilder value = new StringBuilder(16);
+        value.append("realm user=\"");
+        if (authenticationContext != null && authenticationContext.getUser() != null) {
+            value.append(authenticationContext.getUser().getUserName());
+        }
+        value.append('\"');
+        response.setHeader(AUTH_HEADER_NAME, value.toString());
+        if (isRequestFromScim2Api(response)) { // handles only scim 2.0 API error responses.
+            handleScim2ApiErrorResponse(response, error, e);
+        } else if (isRequestFromDCREndpoint(response)) {
+            handleDCRApiErrorResponse(response, error, e);
+        } else {
+            handleErrorResponseForCommonAPIs(response, error, e);
+        }
+    }
+
+
+    private static boolean isRequestFromScim2Api(Response response) {
+
+        if (response.getRequest() == null) {
+            return false;
+        }
+        return response.getRequest().getRequestURI().contains("/scim2");
```

Comment:
can we check starts with ? that will be more accurate

Reply:
This has been fixed with 747c2c448717dcac5a09b0abec9838c5d61ac644